### ISSUE TITLE
Document watermark logic for running migrations

### ISSUE THREAD

**Issue by rjoonas** [guidesmiths/marv#45]
I had a naive assumption that marv would run all new migrations that haven't been recorded in the `migrations` table and was somewhat surprised when it silently skipped migrations with a lower level than what was previously run:

https://github.com/guidesmiths/marv/blob/3a952e7e23a316c41dd851f7f1fdaa0554ba2a57/lib/migrate.js#L68-L71

It’s quite easy to run into this situation when using multiple VCS branches:

* Assume the current migration level is 1.
* Work on two features in parallel, feature A adds a migration with level 2 and feature B adds a migration with level 3.
* Merge and deploy feature B. marv’s migration level becomes 3.
* Merge and deploy feature A. marv’s migration level is 3 so it silently skips migration 2.

Errors like this may be easy to notice in this trivial example but less so if you have multiple contributors working on several branches and if you use less readable serials like UNIX timestamps for indexing (which would otherwise work well to avoid having to manually agree on migration serials between unrelated branches).

I wonder if this logic would be worth mentioning in the README? It could also be nice to have some kind of warning when migrations are skipped, or even better, have this behavior configurable somehow.

**Comment by cressie176** [comment]
Hi @rjoonas,

Sorry you were caught out by this, and I agree it should have been documented. I'll update the README accordingly. I'll also check how easy it would be to add a warning, and a "backfill" option

**Comment by cressie176** [comment]
I had another thought yesterday evening...

Marv already reports duplicate levels, so one way to prevent this is to treat each branch in isolation.

So in your example, both the feature A and feature B branches would add level 2. The first to merge to main/master would be ok, but when the second was merged, marv would fail to run the migrations, hopefully causing the build to break until the migration was renamed with level 3.

I still agree that the README should be improved and that getting marv to report ignored migrations would be a useful feature. I'm also not ruling out the backfill option, just proposing a modified workflow, where branches are considered independent from one another.

**Comment by cressie176** [comment]
I haven't forgotten, just decided to do some maintenance on marv-pg-driver first.

**Comment by cressie176** [comment]
I've documented the behaviour and changed marv to detect and report missed migrations.

After some consideration, I've decided not to support a backfill option. Attempting to pre-empt what the sequence of migration files from multiple branches isn't a scalable process. You won't always know how many migration files you'll need per feature, so the most sensible thing (IMHO) is to treat the branches as isolated, and to continue numbering from the latest level in trunk.

This does mean that when the branches are merged back to trunk there will be duplicate levels, but marv already reports these, and merging back seems like the right place to renumber.
